clash-lib/src/Clash/Core/Util.hs

```diff
@@ -214,35 +214,30 @@ extractTElems supply inScope lrCon brCon resTy s maxN tree =
      -> (Supply,InScopeSet)
      -> Term
      -> ((Supply,InScopeSet),([Term],[(Id, Term)]))
-  go 0 _ ks uniqs0 e = (uniqs1,([elNVar],[(elNId, rhs)]))
-   where
-    tys          = [LitTy (NumTy 0),resTy]
-    (Just idTys) = dataConInstArgTys lrCon tys
+  go 0 _ ks uniqs0 e = fromMaybe (error "extractTElems: failed to project elements") $ do
+    let tys = [LitTy (NumTy 0),resTy]
+    idTys <- dataConInstArgTys lrCon tys
 
-    (uniqs1,[elNId,co,el]) =
+    (uniqs1,[elNId,co,el]) <- pure $
```

Comment:
Ah, because failed monadic pattern bindings desugar to `fail :: MonadFail m => String -> m a`, i.e. it's part of the semantics of the monadic. Whereas regular failed pattern bindings desugar to  `error :: String -> a` and are thus somewhat "extra"-semantical if I had to make up a word.

Reply:
aha!